cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -65,6 +65,133 @@ AzureOptions::AzureOptions() = default;
 
 AzureOptions::~AzureOptions() = default;
 
+Result<AzureOptions> AzureOptions::FromUri(const arrow::internal::Uri& uri,
+                                           std::string* out_path) {
+  AzureOptions options;
+  const auto host = uri.host();
+  std::string container;
+  std::string path;
+  if (arrow::internal::EndsWith(host, options.blob_storage_authority)) {
+    options.account_name =
+        host.substr(0, host.size() - options.blob_storage_authority.size());
+    auto components = internal::SplitAbstractPath(uri.path());
+    if (!components.empty()) {
+      container = components[0];
+      path = internal::JoinAbstractPath(components.begin() + 1, components.end());
+    }
+  } else if (arrow::internal::EndsWith(host, options.dfs_storage_authority)) {
+    options.account_name =
+        host.substr(0, host.size() - options.dfs_storage_authority.size());
+    container = uri.username();
+    path = uri.path();
+  } else {
+    options.account_name = uri.username();
+    std::string host_port = host;
+    const auto port_text = uri.port_text();
+    if (!port_text.empty()) {
+      host_port += ":" + port_text;
+    }
+    options.blob_storage_authority = host_port;
+    options.dfs_storage_authority = host_port;
+    if (uri.scheme() == "abfs") {
+      options.blob_storage_scheme = "http";
+      options.dfs_storage_scheme = "http";
```

Comment:
So let's figure out a way to infer we're talking to Azurite, but not assume that people using `abfs://` are OK with plain-text communications and would remember to intentionally opt-in to security by using `abfss://`. Security should be opt-out, not opt-in. If someone ever complains about `https://` by default we should spec a URI query parameter to disable TLS instead of having it subtly disabled by the scheme chosen.

Even the Microsoft documentation says that they will use TLS even when `abfs` is used in the Haddop driver.

Reply:
Sounds good to me :+1:

Reply:
> We should process abfs and abfss as the same schemes
> We should always use HTTPS for abfs and abfss

Yes.

> We should add a query parameter to use HTTP (scheme=http?)

Probably something more explicit like `?disable_tls`. I would be OK with a change in `AzureOptions` as well. Instead of `scheme` a `bool disable_tls` and then we compute the scheme (http or https) according to it.

Reply:
OK. Let's choose this approach.

I can understand `disable_tls` naming (it implies that TLS is enabled by default) but can we avoid using negative option name (`disable_XXX`)? It's a bit difficult to interpret when we use `if (!disable_XXX)` (double NOTs). Can we use `enable_tls=false` (the default value is `true` of cause)?

Reply:
Yeah. That's better.